python/lsst/daf/butler/remote_butler/server/_gafaelfawr.py

```diff
@@ -0,0 +1,127 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+from collections import defaultdict
+
+import httpx
+import pydantic
+
+from .._authentication import get_authentication_headers
+
+
+class GafaelfawrClient:
+    """REST client for retrieving authentication information from
+    Gafaelfawr.
+
+    Parameters
+    ----------
+    base_url : `str`
+        The top-level HTTP path where Gafaelfawr can be found (e.g.
+        ``"https://data-int.lsst.cloud"``).
+    transport : ``httpx.AsyncBaseTransport``, optional
+        Override the HTTP client's transport.  (For unit tests).
+    """
+
+    def __init__(self, base_url: str, *, transport: httpx.AsyncBaseTransport | None = None) -> None:
+        if transport is None:
+            transport = httpx.AsyncHTTPTransport(retries=3)
+        self._client = httpx.AsyncClient(base_url=base_url, transport=transport, timeout=20.0)
+
+    async def get_groups(self, user_token: str) -> list[str]:
+        response = await self._client.get(
+            "/auth/api/v1/user-info", headers=get_authentication_headers(user_token)
```

Comment:
It's probably slightly more future-proof to move `/auth` here into the base URL. I think the `/auth` prefix will only be used when everything is running in the same domain. The long-term intent is to have Gafaelfawr run under auth.data.lsst.cloud (for example), and in that configuration we'll probably drop the `/auth` prefix.